### ISSUE TITLE
Defect in `gradients` implementation

### ISSUE THREAD

**Issue by mspandit** [malmaud/TensorFlow.jl#410]
There seems to be a defect in the implementation of `gradients`. The same simple program converges in Python, but diverges in Julia:
```
import sys
from tensorflow.examples.tutorials.mnist import input_data
import tensorflow as tf
import numpy

if __name__ == '__main__':
    learning_rate = 0.1
    mini_batch_size = 128
    # Import data
    mnist = input_data.read_data_sets("/tmp/mnist_data", one_hot=True)

    # Create the model
    example_input = tf.placeholder(tf.float32, [None, 784])
    example_output = tf.placeholder(tf.float32, [None, 10])

    def get_feeds():
        batch_inputs, batch_outputs = mnist.train.next_batch(mini_batch_size)
        return {
            example_input: batch_inputs,
            example_output: batch_outputs
        }

    thetas = tf.Variable(tf.zeros([784, 10]))
    theta0 = tf.Variable(tf.zeros([10]))

    model_output = tf.nn.softmax(tf.matmul(example_input, thetas) + theta0)
    cost = tf.reduce_mean(
        -tf.reduce_sum(tf.multiply(example_output, tf.log(model_output)), reduction_indices=1))

    calc_grad_thetas = tf.gradients(xs=[thetas], ys=cost)[0]
    calc_grad_theta0 = tf.gradients(xs=[theta0], ys=cost)[0]
    update_thetas = thetas.assign(thetas - tf.multiply(learning_rate, calc_grad_thetas))
    update_theta0 = theta0.assign(theta0 - tf.multiply(learning_rate, calc_grad_theta0))

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())

        feeds = get_feeds()
        for _ in range(2):
            mo, c, gt, gt0, new_thetas, new_theta0 = sess.run([model_output, cost, calc_grad_thetas, calc_grad_theta0, update_thetas, update_theta0], feed_dict=feeds)
            print("norm(model_output): %0.2f" % numpy.linalg.norm(mo))
            print("cost: %0.2f" % c)
            print("norm(calc_grad_thetas): %0.2f" % numpy.linalg.norm(gt))
            print("norm(calc_grad_theta0): %s" % numpy.linalg.norm(gt0))
            print("norm(update_thetas): %0.2f" % numpy.linalg.norm(new_thetas))

        accuracy = tf.reduce_mean(
            tf.cast(
                tf.equal(tf.argmax(model_output, 1), tf.argmax(example_output, 1)),
                tf.float32))
        print("Accuracy: %0.2f %%" % (100.0 * sess.run(
            accuracy,
            feed_dict={
                example_input: mnist.test.images,
                example_output: mnist.test.labels})))
```
produces the following output:
```
norm(model_output): 3.58
cost: 2.30
norm(calc_grad_thetas): 1.28
norm(calc_grad_theta0): 0.097452894
norm(update_thetas): 0.13
norm(model_output): 3.63
cost: 2.15
norm(calc_grad_thetas): 1.12
norm(calc_grad_theta0): 0.06459431
norm(update_thetas): 0.24
Accuracy: 18.99 %
```
However, the equivalent Julia program
```
using TensorFlow
using MNIST

mutable struct DataLoader
    cur_id::Int
    order::Vector{Int}
end

DataLoader() = DataLoader(1, shuffle(1:60000))

function next_batch(loader::DataLoader, batch_size)
    x = zeros(Float32, batch_size, 784)
    y = zeros(Float32, batch_size, 10)
    for i in 1:batch_size
        x[i, :] = trainfeatures(loader.order[loader.cur_id])
        label = trainlabel(loader.order[loader.cur_id])
        y[i, Int(label)+1] = 1.0
        loader.cur_id += 1
        if loader.cur_id > 60000
            loader.cur_id = 1
        end
    end
    x, y
end

function load_test_set(N=10000)
    x = zeros(Float32, N, 784)
    y = zeros(Float32, N, 10)
    for i in 1:N
        x[i, :] = testfeatures(i)
        label = testlabel(i)
        y[i, Int(label)+1] = 1.0
    end
    x,y
end

learningRate = 0.1
miniBatchSize = 128

# Import data
mnist = DataLoader()

# Create the model
exampleInput = TensorFlow.placeholder(Float64, shape=[nothing, 784])
exampleOutput = TensorFlow.placeholder(Float64, shape=[nothing, 10])

function get_feeds()
  batch_inputs, batch_outputs = next_batch(mnist, miniBatchSize)
  return Dict(
    exampleInput => batch_inputs,
    exampleOutput => batch_outputs
  )
end

thetas = TensorFlow.Variable(TensorFlow.zeros((784, 10)))
theta0 = TensorFlow.Variable(TensorFlow.zeros((1, 10)))

modelOutput = TensorFlow.nn.softmax(
  TensorFlow.matmul(exampleInput, thetas) + theta0)
cost = TensorFlow.reduce_mean(
  -TensorFlow.reduce_sum(
    TensorFlow.multiply(exampleOutput, TensorFlow.log(modelOutput)), axis=2))

calcGradThetas = TensorFlow.gradients(cost, [thetas])[1]
calcGradTheta0 = TensorFlow.gradients(cost, [theta0])[1] # <=== Defect?
updateThetas = TensorFlow.assign(thetas, thetas - TensorFlow.multiply(learningRate, calcGradThetas))
updateTheta0 = TensorFlow.assign(theta0, theta0 - TensorFlow.multiply(learningRate, calcGradTheta0))

sess = TensorFlow.Session()
run(sess, TensorFlow.global_variables_initializer())

feeds = get_feeds()
for i = 1:2
  mo, c, gt, gt0, newThetas, newTheta0 = run(sess, [modelOutput, cost, calcGradThetas, calcGradTheta0, updateThetas, updateTheta0], feeds)
  println("norm(modelOutput): $(norm(mo))")
  println("cost: $(c)")
  println("norm(calcGradThetas): $(norm(gt))")
  println("norm(calcGradTheta0): $(norm(gt0))")
  println("norm(updateThetas): $(norm(newThetas))")
end

accuracy = TensorFlow.reduce_mean(
  TensorFlow.cast(
    TensorFlow.equal(TensorFlow.arg_max(modelOutput, 2), TensorFlow.arg_max(exampleOutput, 2)),
    TensorFlow.Float32))
test_inputs, test_outputs = load_test_set()
println("Accuracy: $(@sprintf("%0.2f", 100.0 * run(sess, accuracy, Dict(
  exampleInput => test_inputs,
  exampleOutput => test_outputs
))))%")
```
produces the following output:
```
norm(modelOutput): 3.577708763999664
cost: 2.302585092994046
norm(calcGradThetas): 237.2033447055505
norm(calcGradTheta0): 0.08103939049931223
norm(updateThetas): 23.720334470555056
norm(modelOutput): 8.485281374238571
cost: NaN
norm(calcGradThetas): NaN
norm(calcGradTheta0): NaN
norm(updateThetas): NaN
Accuracy: 9.80%
```

**Comment by oxinabox** [comment]
Can you cut this down to a MWE?
Ideally without any data loading etc.

**Comment by oxinabox** [comment]
This is unrelated.

Your custom compiled LibTensorFlow is `v1.10.0` and the Python is `v1.10.1`
I suggest you should compile `v1.10.1` libtensorflow

**Comment by colbec** [comment]
Thanks, got it. I forgot to do the ./configure before recompile of tensorflow. Tests now pass.
